File: src-winui/src/GSBT.Core/Services/GameScanPostProcessor.cs
using GSBT.Core.Models;
using System.Globalization;

namespace GSBT.Core.Services;

/// <summary>
/// Post-scan merge: multiple Steam library rows that resolve to the same save data (exact path, or nested folders under the same franchise title).
/// </summary>
public static class GameScanPostProcessor
{
    /// <summary>
    /// Merges duplicate save rows: same resolved folder, or (Steam only) same franchise prefix in the title with overlapping save paths on disk.
    /// </summary>
    public static (IReadOnlyList<SaveScanResult> Kept, IReadOnlyList<string> DroppedCatalogNames) DeduplicateBySharedSaveRoot(
        IReadOnlyList<SaveScanResult> results)
    {
        if (results.Count == 0)
        {
            return (results, []);
        }

        var n = results.Count;
        var parent = new int[n];
        for (var i = 0; i < n; i++)
        {
            parent[i] = i;
        }

        int Find(int i)
        {
            while (parent[i] != i)
            {
                parent[i] = parent[parent[i]];
                i = parent[i];
            }

            return i;
        }

        void Union(int a, int b)
        {
            var ra = Find(a);
            var rb = Find(b);
            if (ra != rb)
            {
                parent[rb] = ra;
            }
        }

        for (var i = 0; i < n; i++)
        {
            for (var j = i + 1; j < n; j++)
            {
                var ki = NormalizeDiskSaveKey(results[i]);
                var kj = NormalizeDiskSaveKey(results[j]);
                if (ki is not null && kj is not null && string.Equals(ki, kj, StringComparison.OrdinalIgnoreCase))
                {
                    Union(i, j);
                    continue;
                }

                if (ki is null || kj is null)
                {
                    continue;
                }

                if (!IsSteam(results[i]) || !IsSteam(results[j]))
                {
                    continue;
                }

                if (!string.Equals(FranchiseTitle(results[i]), FranchiseTitle(results[j]), StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                if (PathsOverlap(results[i].SavePathResolved!, results[j].SavePathResolved!))
                {
                    Union(i, j);
                }
            }
        }

        var byRoot = new Dictionary<int, List<int>>();
        for (var i = 0; i < n; i++)
        {
            var r = Find(i);
            if (!byRoot.TryGetValue(r, out var list))
            {
                list = [];
                byRoot[r] = list;
            }

            list.Add(i);
        }

        var dropped = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var winnerRowIdByRoot = new Dictionary<int, string>();
        foreach (var kv in byRoot)
        {
            var indices = kv.Value;
            var members = indices.Select(ix => results[ix]).ToList();
            var w = PickPreferredSameSaveRoot(members);
            winnerRowIdByRoot[kv.Key] = w.RowId;
            foreach (var ix in indices)
            {
                if (!string.Equals(results[ix].RowId, w.RowId, StringComparison.OrdinalIgnoreCase))
                {
                    dropped.Add(results[ix].Name);
                }
            }
        }

        var kept = new List<SaveScanResult>();
        for (var i = 0; i < n; i++)
        {
            var root = Find(i);
            if (string.Equals(results[i].RowId, winnerRowIdByRoot[root], StringComparison.OrdinalIgnoreCase))
            {
                kept.Add(results[i]);
            }
        }

        return (kept, dropped.ToList());
    }

    private static string? NormalizeDiskSaveKey(SaveScanResult r)
    {
        if (string.IsNullOrWhiteSpace(r.SavePathResolved))
        {
            return null;
        }

        if (r.SaveInRegistryOnly)
        {
            return null;
        }

        try
        {
            return Path.GetFullPath(r.SavePathResolved).TrimEnd('\\', '/').ToLowerInvariant();
        }
        catch
        {
            return null;
        }
    }

    private static bool IsSteam(SaveScanResult r) =>
        string.Equals(r.Platform, "Steam", StringComparison.OrdinalIgnoreCase);

    /// <summary>First segment before ':' (Relic-style expansions), otherwise full name.</summary>
    private static string FranchiseTitle(SaveScanResult r)
    {
        var n = r.Name;
        var c = n.IndexOf(':');
        return (c < 0 ? n : n[..c]).Trim();
    }

    private static bool PathsOverlap(string a, string b)
    {
        try
        {
            var fa = Path.GetFullPath(a).TrimEnd('\\', '/');
            var fb = Path.GetFullPath(b).TrimEnd('\\', '/');
            if (string.Equals(fa, fb, StringComparison.OrdinalIgnoreCase))
            {
                return true;
            }

            return fa.StartsWith(fb + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
                || fa.StartsWith(fb + Path.AltDirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
                || fb.StartsWith(fa + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
                || fb.StartsWith(fa + Path.AltDirectorySeparatorChar, StringComparison.OrdinalIgnoreCase);
        }
        catch
        {
            return false;
        }
    }

    private static SaveScanResult PickPreferredSameSaveRoot(IReadOnlyList<SaveScanResult> group)
    {
        if (group.Count == 1)
        {
            return group[0];
        }

        foreach (var candidate in group.OrderBy(x => x.Name.Length).ThenBy(x => x.Name, StringComparer.OrdinalIgnoreCase))
        {
            var n = candidate.Name;
            if (group.All(o =>
                    string.Equals(o.Name, n, StringComparison.OrdinalIgnoreCase)
                    || o.Name.StartsWith(n + ":", StringComparison.OrdinalIgnoreCase)))
            {
                return candidate;
            }
        }

        SaveScanResult? best = null;
        var bestId = long.MaxValue;
        foreach (var o in group)
        {
            if (!long.TryParse((o.AppId ?? string.Empty).Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out var id))
            {
                continue;
            }

            if (id < bestId)
            {
                bestId = id;
                best = o;
            }
        }

        if (best is not null)
        {
            return best;
        }

        return group.OrderBy(x => x.Name, StringComparer.OrdinalIgnoreCase).First();
    }
}

File: src-pyqt/core/game_scan_post_processor.py
"""
Post-scan merge: collapse rows that share the same on-disk save folder (DLC / franchise duplicates).
"""

from __future__ import annotations

import os
from typing import Any, Dict, List, Tuple


def _normalize_disk_save_key(row: Dict[str, Any]) -> str | None:
    resolved = row.get("save_path_resolved")
    if not resolved or row.get("save_in_registry_only"):
        return None
    try:
        return os.path.normpath(str(resolved)).rstrip("\\/").lower()
    except (OSError, ValueError):
        return None


def _is_steam(row: Dict[str, Any]) -> bool:
    return str(row.get("platform") or "").lower() == "steam"


def _franchise_title(row: Dict[str, Any]) -> str:
    name = str(row.get("name") or "")
    idx = name.find(":")
    return (name if idx < 0 else name[:idx]).strip()


def _paths_overlap(a: str, b: str) -> bool:
    try:
        fa = os.path.normpath(a).rstrip("\\/")
        fb = os.path.normpath(b).rstrip("\\/")
        if fa.lower() == fb.lower():
            return True
        sep = os.sep
        return (
            fa.lower().startswith(fb.lower() + sep)
            or fb.lower().startswith(fa.lower() + sep)
        )
    except (OSError, ValueError):
        return False


def _pick_preferred_same_save_root(group: List[Dict[str, Any]]) -> Dict[str, Any]:
    if len(group) == 1:
        return group[0]

    for candidate in sorted(group, key=lambda x: (len(str(x.get("name") or "")), str(x.get("name") or "").lower())):
        name = str(candidate.get("name") or "")
        if all(
            str(o.get("name") or "").lower() == name.lower()
            or str(o.get("name") or "").lower().startswith(name.lower() + ":")
            for o in group
        ):
            return candidate

    best: Dict[str, Any] | None = None
    best_id = float("inf")
    for row in group:
        app_id = str(row.get("app_id") or "").strip()
        try:
            gid = int(app_id)
        except ValueError:
            continue
        if gid < best_id:
            best_id = gid
            best = row
    if best is not None:
        return best

    return sorted(group, key=lambda x: str(x.get("name") or "").lower())[0]


def deduplicate_by_shared_save_root(
    results: List[Dict[str, Any]],
) -> Tuple[List[Dict[str, Any]], List[str]]:
    """Return (kept rows, dropped catalog names)."""
    n = len(results)
    if n == 0:
        return results, []

    parent = list(range(n))

    def find(i: int) -> int:
        while parent[i] != i:
            parent[i] = parent[parent[i]]
            i = parent[i]
        return i

    def union(a: int, b: int) -> None:
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[rb] = ra

    for i in range(n):
        for j in range(i + 1, n):
            ki = _normalize_disk_save_key(results[i])
            kj = _normalize_disk_save_key(results[j])
            if ki and kj and ki == kj:
                union(i, j)
                continue
            if not ki or not kj:
                continue
            if not _is_steam(results[i]) or not _is_steam(results[j]):
                continue
            if _franchise_title(results[i]).lower() != _franchise_title(results[j]).lower():
                continue
            ra = results[i].get("save_path_resolved")
            rb = results[j].get("save_path_resolved")
            if ra and rb and _paths_overlap(str(ra), str(rb)):
                union(i, j)

    by_root: Dict[int, List[int]] = {}
    for i in range(n):
        root = find(i)
        by_root.setdefault(root, []).append(i)

    dropped: set[str] = set()
    winner_row_id_by_root: Dict[int, str] = {}
    for root, indices in by_root.items():
        members = [results[ix] for ix in indices]
        winner = _pick_preferred_same_save_root(members)
        winner_row_id = str(winner.get("row_id") or "")
        winner_row_id_by_root[root] = winner_row_id
        for ix in indices:
            row_id = str(results[ix].get("row_id") or "")
            if row_id.lower() != winner_row_id.lower():
                dropped.add(str(results[ix].get("name") or ""))

    kept: List[Dict[str, Any]] = []
    for i in range(n):
        root = find(i)
        row_id = str(results[i].get("row_id") or "")
        if row_id.lower() == winner_row_id_by_root[root].lower():
            kept.append(results[i])

    return kept, [name for name in dropped if name]
